src/common/include/parallel_macros.fpp

```diff
@@ -0,0 +1,404 @@
+        #:set data_dependency_val = data_dependency 
+    #:else
+        #:set data_dependency_val = ''
+    #:endif
+    
+    #:set private_val = GEN_PRIVATE_STR(private, False)
+
+    #:set reduction_val = GEN_REDUCTION_STR(reduction, reductionOp)
+
+    #:set extraAccArgs_val = GEN_EXTRA_ARGS_STR(extraAccArgs)
+
+    #:set clause_val = collapse_val.strip('\n') + parallelism_val.strip('\n') + &
+        & data_dependency_val.strip('\n') + private_val.strip('\n') + &
+        & reduction_val.strip('\n')
+    #:set acc_directive = '!$acc loop ' + &
+        & clause_val + extraAccArgs_val.strip('\n')
+    $:acc_directive
+#:enddef
+
+#:def GPU_DATA(copy=None, copyin=None, copyinReadOnly=None, copyout=None, create=None, no_create=None, present=None, deviceptr=None, attach=None, default=None, extraAccArgs=None)
+    
+    #:set copy_val = GEN_COPY_STR(copy)
+
+    #:set copyin_val = GEN_COPYIN_STR(copyin, False).strip('\n') + GEN_COPYIN_STR(copyinReadOnly, True).strip('\n')
+
+    #:set copyout_val = GEN_COPYOUT_STR(copyout)
+
+    #:set create_val = GEN_CREATE_STR(create)
+
+    #:set no_create_val = GEN_NOCREATE_STR(no_create)
+
+    #:set present_val = GEN_PRESENT_STR(present)
+
+    #:set deviceptr_val = GEN_DEVICEPTR_STR(deviceptr)
+
+    #:set attach_val = GEN_ATTACH_STR(attach)
+
+    #:set default_val = GEN_DEFAULT_STR(default)
+
+    #:set extraAccArgs_val = GEN_EXTRA_ARGS_STR(extraAccArgs)
+
+    #:set clause_val = copy_val.strip('\n') + copyin_val.strip('\n') + &
+        & copyout_val.strip('\n') + create_val.strip('\n') + &
+        & no_create_val.strip('\n') + present_val.strip('\n') + & 
+        & deviceptr_val.strip('\n') + attach_val.strip('\n') + &
+        & default_val.strip('\n')
+    #:set acc_directive = '!$acc data ' + clause_val + extraAccArgs_val.strip('\n')
+    $:acc_directive
+#:enddef
+
+#:def GPU_ENTER_DATA(copyin=None, copyinReadOnly=None, create=None, attach=None, extraAccArgs=None)
+    #:set copyin_val = GEN_COPYIN_STR(copyin, False).strip('\n') + GEN_COPYIN_STR(copyinReadOnly, True).strip('\n')
+
+    #:set create_val = GEN_CREATE_STR(create)
+
+    #:set attach_val = GEN_ATTACH_STR(attach)
+
+    #:set extraAccArgs_val = GEN_EXTRA_ARGS_STR(extraAccArgs)
+
+    #:set clause_val = copyin_val.strip('\n') + create_val.strip('\n') + attach_val.strip('\n')
+    #:set acc_directive = '!$acc enter data ' + clause_val + extraAccArgs_val.strip('\n')
+    $:acc_directive
+#:enddef
+
+#:def GPU_EXIT_DATA(copyout=None, delete=None, detach=None, extraAccArgs=None)
+    #:set copyout_val = GEN_COPYOUT_STR(copyout)
+
+    #:set delete_val = GEN_DELETE_STR(delete)
+
+    #:set detach_val = GEN_DETACH_STR(detach)
+
+    #:set extraAccArgs_val = GEN_EXTRA_ARGS_STR(extraAccArgs)
+
+    #:set clause_val = copyout_val.strip('\n') + delete_val.strip('\n') + detach_val.strip('\n')
+    #:set acc_directive = '!$acc exit data ' + clause_val + extraAccArgs_val.strip('\n')
```

Comment:
just using this macro as a random example. do consider how this will work when openmp is added. from my view it looks like the end of each macro will need a few lines of if statements checking to see if it's going to output OMP or ACC. i suppose another option is that it writes both acc and omp directives (then compiler chooses its own route), which would only be two extra lines of code i think (for each macro, though), `:set omp_directive = '!$omp <...>'(\n)` then something like `set directive = acc directive + omp directive`